File: TypeScript/src/data-structures/lists/entities/LinkedList.ts
import { List } from '../lib/List'

import { Node } from '../../models/Node'

import { CallbackFilterFn, CallbackMapFn } from './../../../util/types'
import { concatLists, filterList, mapList } from './../../../util/functions'

export class LinkedList<T> extends List<T> {
  push(el: T) {
    const node = new Node(el)

    if (!this._head) {
      this._head = node
    } else {
      this._tail.next = node
    }

    this._tail = node

    this._length++
    return this._length
  }

  insert(el: T, index = 0) {
    if (index >= 0 && index <= this._length) {
      const node = new Node(el)

      if (!index) {
      
        if (!this._head) {
          this._head = node
          this._tail = node
      
        } else {
          node.next = this._head
          this._head = node
        }
      
      } else if (index === this._length) {
        this._tail.next = node
        this._tail = node
      
      } else {
        const ancestor = this.getNode(index - 1)
        const current = ancestor.next
        node.next = current
        ancestor.next = node
      }

      this._length++
      return this._length
    }
  }

  removeAt(index: number) {
    if (this._head && index >= 0 && index < this._length) {
      let pointer = this._head

      if (!index) {
        this._head = this._head.next

        if (!this._head) {
          this._tail = null
        }
      
      } else {
        const ancestor = this.getNode(index - 1)
        pointer = ancestor.next
        ancestor.next = pointer.next
        this._tail = ancestor
      }

      this._length--
      return pointer.el
    }
  }

  toString() {
    let pointer = this._head

    if (!pointer) {
      return ''
    }

    let str = `${pointer.el}`
    pointer = pointer.next

    for (let i = 1; i < this._length && pointer; i++) {
      str = `${str} -> ${pointer.el}`
      pointer = pointer.next
    }
    return `${str} ->`
  }

  concat(list: List<T>) {
    const newList = new LinkedList<T>()
    concatLists(this, list, newList)

    return newList
  }

  filter(callbackFn: CallbackFilterFn<T>) {
    const newList = new LinkedList<T>()
    filterList(this, newList, callbackFn)

    return newList
  }

  map(callbackFn: CallbackMapFn<T>) {
    const newList = new LinkedList<T>()
    mapList(this, newList, callbackFn)

    return newList
  }
}

File: Go/data-structures/lists/lib/linked-list/linked-list.go
package linkedList

import (
	"Go/data-structures/models/node"
	"Go/util"
)

type LinkedList struct {
	Head   *node.Node
	Tail   *node.Node
	Length int
}

func (list *LinkedList) Merge(listHead *node.Node, listTail *node.Node, listLength int) {
	list.Head = listHead
	list.Tail = listTail
	list.Length = listLength
}

func (list *LinkedList) GetNode(index int) *node.Node {
	if index >= 0 && index < list.Length {
		pointer := list.Head
		for i := 0; i < index && pointer != nil; i++ {
			pointer = pointer.Next
		}
		return pointer
	}
	return nil
}

func (list *LinkedList) GetElementAt(el any, index int) bool {
	pointer := util.GetNode(list.Head, list.Length, index)
	if pointer != nil {
		el = pointer.El
		return true
	}
	return false
}

func (list *LinkedList) SetElementAt(el any, index int) bool {
	pointer := util.GetNode(list.Head, list.Length, index)
	if pointer != nil {
		pointer.El = el
		return true
	}
	return false
}

func (list *LinkedList) IndexOf(index *int, el any) bool {
	pointer := list.Head
	i := 0
	for pointer != nil {
		if pointer.El == el {
			*index = i
			return true
		}
		pointer = pointer.Next
		i++
	}
	return false
}

func (list *LinkedList) Clear() {
	list.Head = nil
	list.Tail = nil
	list.Length = 0
}
